File: ruby/simple_pong/player.rb
require './constants.rb'

class Player

  def initialize(x, color, up_key, down_key)
    @speed = 10
    @rect = Rectangle.new(x: x, y: 100, width: 25, height: 150, color: color)
    @up_key = up_key
    @down_key = down_key
  end

  def key_held(key)
    case key
    when @up_key
      @rect.y -= @speed
    when @down_key
      @rect.y += @speed
    end

    @rect.y = @rect.y.clamp(0, HEIGHT - 150)
  end

end

File: go/pongiten2/player.go
package main

import (
	"github.com/hajimehoshi/ebiten/v2"
)

type Player struct {
	Img *ebiten.Image
	Opt *ebiten.DrawImageOptions

	PosX, PosY    float64
	Width, Height float64

	UpKey   ebiten.Key
	DownKey ebiten.Key
}

func (p *Player) Update() error {
	if ebiten.IsKeyPressed(p.DownKey) {
		p.PosY += PlayerSpeed
	}

	if ebiten.IsKeyPressed(p.UpKey) {
		p.PosY -= PlayerSpeed
	}

	p.PosY = clamp(p.PosY, 0, WinH-p.Height)

	p.Opt.GeoM.Reset()
	p.Opt.GeoM.Scale(PlayerScale, PlayerScale)
	p.Opt.GeoM.Translate(p.PosX, p.PosY)

	return nil
}

func (p *Player) Reset() {
	p.PosY = float64(WinH-p.Height) / 2
}

func clamp(value float64, min float64, max float64) float64 {
	if value < min {
		return min
	}

	if value > max {
		return max
	}

	return value
}
